scarif/cartridge_accessor.py

```diff
@@ -0,0 +1,174 @@
+            raise SoftLimitExceededException('%s out of range [%s %s]' % (
+                y, self.y_limits[0], self.y_limits[1]))
+        return super(Head, self).goto_x_y(x, y)
+
+    # movement methods
+    def up(self, count):
+        self._set_vertical_acceleration(self.accel_v)
+        curr_stat = self.current_status()
+        x_new = float(curr_stat['w_pos'][0]) + count
+        self.goto_x_y('%.3f' % (x_new), curr_stat['w_pos'][1])
+
+    def left(self, count):
+        self._set_horizontal_acceleration(self.accel_h)
+        curr_stat = self.current_status()
+        y_new = float(curr_stat['w_pos'][1]) + count
+        self.goto_x_y(curr_stat['w_pos'][0], '%.3f' % (y_new))
+
+    def down(self, count):
+        self.up(-1*count)
+
+    def right(self, count):
+        self.left(-1*count)
+
+
+class Picker(object):
+    """A class with methods for controlling a tape library picker. Three
+    methods are implemented:
+        * insert - move a tape from the picker to a slot/drive
+        * retrieve - move a tape from a slot/drive to the picker
+        * home - reset the picker - not elegant, but gets the job done
+
+    Parameters
+    ----------
+    port: string: USB serial device to connect to, e.g. /dev/serial
+    baudrate: integer : communications bit rate, e.g. 115200
+    timeout: integer : USB serial communcations timeout in seconds
+    """
+    def __init__(self, port, baudrate, timeout):
+        super(Picker, self).__init__()
+        self._serial = Serial()
+        self._serial.port = port
+        self._serial.baudrate = baudrate
+        self._serial.timeout = timeout
+        self.open()
+
+    def open(self):
+        """Open the USB serial port to the arduino."""
+        self._serial.open()
+
+    def close(self):
+        """Close the USB serial port to the arduino."""
+        self._serial.close()
+
+    def configure(self):
+        """Configure the picker. Currently does nothing."""
+        pass
+
+    def insert(self):
+        """Execute an insert sequence on the picker arduino."""
+        self._serial.write("i")
+        while True:
+            data = self._serial.readline()
+            if data:
+                logger.debug(data.rstrip('\n'))
+                if data.startswith('end'):
+                    break
+
+    def retrieve(self):
+        """Execute a retrieve sequence on the picker arduino."""
+        self._serial.write("r")
+        while True:
+            data = self._serial.readline()
+            if data:
+                logger.debug(data.rstrip('\n'))
+                if data.startswith('end'):
+                    break
+
+    def home(self):
+        """Execute a home sequence on the picker arduino."""
+        self._serial.write("h")
+        while True:
+            data = self._serial.readline()
+            if data:
+                logger.debug(data.rstrip('\n'))
+                if data.startswith('end'):
+                    break
```

Comment:
Home, Retrieve, and Insert are functionally the same method.
Can we not create a method lets call it _send_command
and Call it from those methods.

e.g.

> def home(self):
>        self._send_command("h")
>
> def _send_command(self, command):
>         """Execute a home sequence on the picker arduino."""
>         self._serial.write(command)
>         while True:
>             data = self._serial.readline()
>             if data:
>                 logger.debug(data.rstrip('\n'))
>                 if data.startswith('end'):
>                     break

Reply:
Thanks :)
I've created:
 self.cmds = {'insert':'i', 'retrieve':'r', 'home':'h'}

Now just call:
self._send_cmd(self.cmds['insert'])